### DISCUSSION TITLE
4.0.0 Beta1

### DISCUSSION THREAD

**Discussion by jc3213** [jc3213/download_with_aria2#45] [Announcements]
- Split `New Download Window` from `Task Manager` as a new browser window
- Added new options for `Context menu` or `Capture Browser Downloads`
    - `Display Download Prompt` to show slimmed `New Download Window` with options
    - `Forward Request Headers` that make the extension not to forward request headers for some sites
- Added support for `Dark Mode`
    - Chromium requires Windows 10 or later
    - Firefox just turn on `Dark` theme
- Optimized UI layout, code logic, and i18n

<hr /><em>This discussion was created from the release <a href='https://github.com/jc3213/download_with_aria2/releases/tag/4.0.0b1'>4.0.0 Beta1</a>.</em>

**Comment by jc3213** [comment]
@ivysrono @wordpure 可以的话帮忙测试一下吧，我自己体验下来没太大问题

**Comment by jc3213** [comment]
- Optimized UI Layout
<hr /><em>This discussion was created from the release <a href='https://github.com/jc3213/download_with_aria2/releases/tag/4.0.0b2'>4.0.0 Beta2</a>.</em><br />
<em>This discussion was created from the release <a href='https://github.com/jc3213/download_with_aria2/releases/tag/4.0.0b3'>4.0.0 Beta3</a>.</em><br />
<em>This discussion was created from the release <a href='https://github.com/jc3213/download_with_aria2/releases/tag/4.0.0b4'>4.0.0 Beta4</a>.</em>

**Comment by jc3213** [comment]
- Fixed `Options` window for Manifest V3
- Shortcuts to open `Options` and `New Download`
- `Ctrl` + `Enter` and `Alt` + `S` to trigger `Submit` button on `New Download` Window
- `Ctrl` + `S` to trigger `Save` button on `Options` Window
- `Ctrl` + `Z` to trigger `Undo` button on `Options` Window
- `Ctrl` + `Y` to trigger `Redo` button on `Options` Window

<hr /><em>This discussion was created from the release <a href='https://github.com/jc3213/download_with_aria2/releases/tag/4.0.0b5'>4.0.0 Beta5</a>.</em>

**Comment by wordpure** [comment]
更新到 4.0.0 Beta5 之后，Firefox 在弹出窗口直接点击「新下载」无响应。

可能是 Firefox 打开新窗口后，扩展程序的弹出页自动关闭，结果 `window.close()` 把下载窗口给关了。

![图片](https://user-images.githubusercontent.com/68013962/201087271-0d8ab3b3-23a9-492e-a828-1ec738ae38fe.png)

**Reply by jc3213** [reply]
这个跟下面那个问题应该是一个毛病，我在4.0.0 Beta5使用`chrome.windows.getCurrent()`来获取当前的窗口，以方便确定弹出新下载窗口的位置，因为manifest v3是没有`window`以及`window.screen`变量的。但是目前不管任何方案，在火狐之下无法打开新下载窗口，而且控制台没有错误报告，就挺离谱的。

**Comment by wordpure** [comment]
同样是设置「总是抓取浏览器下载」，4.0.0.5 不会自动捕获下载，只能通过右键菜单抓取。

不知道跟「选择 API」的消失有没有关系。

![图片](https://user-images.githubusercontent.com/68013962/201108758-602ef954-ceba-4fb1-9244-b43ef3e44672.png)

![图片](https://user-images.githubusercontent.com/68013962/201108768-aa5f5d4c-45e5-4b89-9b37-0cc80e1bee24.png)

**Reply by jc3213** [reply]
4.0.0 Beta5 我还没在Firefox下面测试，准确的说是修改太多测不过来。我在4.0.0 Beta下修复以后就没太在意了。回头看一看

**Reply by jc3213** [reply]
这个问题我这里测试是正常的啊？抓取API也是正常显示的。如果不弹出新下载窗口的话就是正常抓取，应该是弹出下载窗口相关的代码出问题了。但是控制台没有任何错误提示，我很迷茫

**Comment by wordpure** [comment]
另外有几个建议希望能考虑下。

1. 可以给有快捷键的按钮加上相应的 title 属性，不然没看更新说明的用户可能永远不知道有这些快捷键。
	不过这样 title 对应的样式也得排除掉 button 才行。

![图片](https://user-images.githubusercontent.com/68013962/201110954-b3d23b31-e765-4b9f-b81d-a6674186dab5.png)

2. 可以把选项页的修改事件从 change 改为 input，这样文本字段修改后不用触发 blur 事件也可以保存或撤销。
	
	![图片](https://user-images.githubusercontent.com/68013962/201112973-4504cdd3-b614-49e7-a0e5-22c2a6f76b98.png)

3. 可以给选项页加个 onbeforeunload 事件，避免没保存的时候不小心关闭了页面。

4. 下载、选项窗口扩展快捷键对应的文本貌似还没有 i18n 文本。

	![图片](https://user-images.githubusercontent.com/68013962/201116507-9d0bb848-43f7-492d-9d30-0b6c76e1de14.png)

**Reply by jc3213** [reply]
1. 添加title是可行的。其实这里我想用`:has`选择器，但是火狐必须103以上还必须手动在`about:config`里面打开才行，否则报错。
2. input的话计算太多，change计算少点会好点，我这么做只是想减少计算。
3. `onbeforeunload` 不是所有情况下都能正常触发，意义不大，并且不一定是所有人都愿意关闭窗口即保存。
4. 窗口这里的i18n暂时还没做，其实不难就是了。

**Reply by jc3213** [reply]
另外这个扩展本来使用者就不太多，我觉得没啥大必要专门标注页面内部的快捷键，虽然我想完善wiki，但是一时半会还做不到。

**Reply by jc3213** [reply]
4. i18n已经做了。

**Reply by wordpure** [reply]
>     2. input的话计算太多，change计算少点会好点，我这么做只是想减少计算。

其实我觉得表单控件没必要弄「撤销」「恢复」这些的，选项界面修改本来就不多，文本字段又自带这个功能。完全可以只留一个「保存」按钮，简单直观，不容易出错。

**Comment by jc3213** [comment]
大概知道原因了，因为使用了2次`chrome.windows` API，导致代码有延迟，在代码完成执行之前就被`window.close()`关闭掉了（因为chromium不会自动关闭窗口），因此改成`async function`配合`await`就行了。然后我测试抓取啥得也正常了，就挺离谱的。
![ok](https://user-images.githubusercontent.com/8744936/201245481-71d76f04-6ada-4eae-b532-c46b0b8ddbb6.png)
![capture](https://user-images.githubusercontent.com/8744936/201246474-fcfe55ce-13fa-4f9f-ba00-ef4b077b60e9.png)

**Comment by jc3213** [comment]
发现问题了，窗口化模式下的Firefox无法使用`windows.create` API新建窗口，原因未知。并且通过`Promise`版本的`browser.windows.create`测试也没有错误发生，而是明明正确执行了但是却没有新建窗口。

本世纪最大疑问了。最后把Firefox窗口最大化就能解决了。这……只能说这个问题必须让Mozilla来修复，我无能为力。

补传了个视频，这个问题我无法修复 [Firefox窗口化下无法新建窗口.zip](https://github.com/jc3213/download_with_aria2/files/9986528/Firefox.zip)

**Reply by wordpure** [reply]
我这里的情况和你差别很大😂

1. Firefox 窗口未最大化的时候，也可以正常打开新下载窗口。
2. 选项界面的「下载文件夹」字段没有「浏览器」的选项，也不会显示「选择API」字段。
3. 即使设置了「总是抓取浏览器下载」，还是只能通过右键菜单抓取下载。
4. 右键菜单打开的「确认下载信息」窗口，点击「选项」按钮无响应。

https://user-images.githubusercontent.com/68013962/201457310-44d73423-982d-4375-904e-4a9959e24bcb.mp4

扩展版本：main 分支最新代码
浏览器版本：Firefox 107.0b9
操作系统版本：Windows 22H2 22621.819

**Reply by jc3213** [reply]
能看下控制台有没有什么错误？我这个纯粹就是干净的profile，没你那边那么多扩展

**Reply by wordpure** [reply]
>     2. 选项界面的「下载文件夹」字段没有「浏览器」的选项，也不会显示「选择API」字段。
> 
>     3. 即使设置了「总是抓取浏览器下载」，还是只能通过右键菜单抓取下载。

这两点是我的锅，调试扩展的时候用的是 Chromium 文件夹，把 Firefox 的修改加进去后就正常了。

顺带一提，可以给「确认下载信息」和「转发请求头」设个默认值，现在默认空值不好看😂

![图片](https://user-images.githubusercontent.com/68013962/201458611-0a6ade36-ba5d-416a-884a-d22bdbe3b61f.png)

![图片](https://user-images.githubusercontent.com/68013962/201458612-51920a0d-32aa-4985-97e9-c3509a235582.png)

**Reply by wordpure** [reply]
>     4. 右键菜单打开的「确认下载信息」窗口，点击「选项」按钮无响应。

这个控制台没有报错信息。

**Reply by jc3213** [reply]
我这边测试全新安装是正常的啊？你把整个源码全部重新下过试试看？
![ass](https://user-images.githubusercontent.com/8744936/201459105-af8b084f-10a9-46c3-8734-972cd832b0d5.png)

**Reply by jc3213** [reply]
我这边测试全新安装是正常的啊？你把整个源码全部重新下过试试看？
![ass](https://user-images.githubusercontent.com/8744936/201459105-af8b084f-10a9-46c3-8734-972cd832b0d5.png)

**Reply by jc3213** [reply]
然后关于弹窗不正常，多半是因为原生火狐的窗口低于某个分辨率以后，新建窗口跟原生窗口形成某种比例所以无法生成，目前测试这个阈值应该是1300左右的宽度，低了就不新建窗口。

**Reply by wordpure** [reply]
> 我这边测试全新安装是正常的啊？你把整个源码全部重新下过试试看？

在 Firefox 的新配置文件安装的话，这两个字段就正常显示了，应该是我现在用的配置文件里安装的 3.21.7 里的设置干扰到了。

**Reply by jc3213** [reply]
应该是默认的话因为没做任何override，这两个就是默认为空白。虽然可以写override但是感觉没必要……。然后`新下载`窗口的`选项`按钮不工作我不知道你那边为什么，我这边正常

然后我测试1920x1080的屏幕，默认火狐窗口位置top 4px, left 4px。如果窗口宽度小于width 1374px，那么就不会弹出窗口，刚好1374px都能正常显示
![1374px](https://user-images.githubusercontent.com/8744936/201459875-32cb2e88-aeff-4972-be1c-1a87f0bd41c0.png)

**Reply by wordpure** [reply]
> 然后关于弹窗不正常，多半是因为原生火狐的窗口低于某个分辨率以后，新建窗口跟原生窗口形成某种比例所以无法生成，目前测试这个阈值应该是1300左右的宽度，低了就不新建窗口。

这个应该是浏览器的安全限制，Chrome 也有，把Chrome 的窗口再调小一些就会出现。

**Reply by jc3213** [reply]
应该是新建窗口popup类型的窗口不能超过现在窗口的范围。

目前看起来chromium的算法更合理。火狐这个通过windows.getCurrent()获取到的窗口跟实际触发时的窗口大小不一样。更接近于window.screen下面的数值，导致新弹窗的位置太偏右，算出来低于1374就超出边界。

目前你用新档案测试其他功能有问题吗？我自己能测到的太有限了。唉，很多东西还想改。

**Reply by wordpure** [reply]
> 然后`新下载`窗口的`选项`按钮不工作我不知道你那边为什么，我这边正常

刚才又试了下，`选项`按钮其实是正常工作的，点击后可以切换选项相关的字段是否出现。

只是点击后窗口大小没有改变，新出现的内容又不在可视区域，导致我误以为点击没反应而已🤣

不过，点击后前进到选项页面，保存或取消后再返回下载页面会不会更合理？

一方面更直观，另一方面修改设置的时候不会受下载倒计时的影响。

**Reply by jc3213** [reply]
这里其实我想改变的，但是实在不太会写逻辑，虽然好像用windows.update就能改写窗口大小就是了，同时清除掉倒计时。

> 不过，点击后前进到选项页面，保存或取消后再返回下载页面会不会更合理？

但是第二点我没太理解？能否向下说明下。另外推荐下面新开个窗口，这个讨论太长导致折叠了。

**Reply by wordpure** [reply]
> 目前你用新档案测试其他功能有问题吗？我自己能测到的太有限了。唉，很多东西还想改。

有两个奇怪的地方，但不清楚是不是这次更新带来的。

1. 直接点击下载链接，弹出的下载确认窗口可以获取到文件名，通过右键菜单弹出的不行。

![图片](https://user-images.githubusercontent.com/68013962/201460769-c740e570-6465-4dd7-8594-7261e94baeb5.png)

![图片](https://user-images.githubusercontent.com/68013962/201460772-de535c2e-35c3-4d79-a3d0-a5987ededeb2.png)

2. 把选项界面的「刷新间隔」从 `1` 改为 `5` ，点击「撤销」，刷新间隔就变成了 `5000`，刷新页面后才恢复正常。
	另外「刷新间隔」虽然设置了最大值为 `10`，但可以把值改为 `11` 后正常保存。
	点击「保存」的时候也不会校验数据的有效性。比方说，把刷新间隔从原来的 `5` 改成 `xx`，点击「保存」后显示正常，但刷新界面后就变成 `0` 了。

**Comment by wordpure** [comment]
>    > 不过，点击后前进到选项页面，保存或取消后再返回下载页面会不会更合理？
>
> 但是第二点我没太理解？能否向下说明下。

类似下图这种效果。

![图片](https://user-images.githubusercontent.com/68013962/201461192-94178049-8a69-4fab-8744-b8680fd43211.png)

![图片](https://user-images.githubusercontent.com/68013962/201461195-00ee117c-100c-4843-bd53-5a2773743b0f.png)

**Reply by jc3213** [reply]
这样呢？
![before](https://user-images.githubusercontent.com/8744936/201461416-1d14a7e4-d67e-4aff-b696-d4e2596bf20c.png)
![after](https://user-images.githubusercontent.com/8744936/201461420-614e1b59-a515-4292-912b-200709e36ef2.png)

**Reply by wordpure** [reply]
这种也行，就是有点反直觉。

或许可以把选项按钮改成 `summary` ，同样是点击展开，再次点击收起，这种符合直觉多了。

**Reply by jc3213** [reply]
我现在主要是想默认打开新下载就包含了所有选项，而抓取下载这种则是个简化窗口，但是用户能通过点击按钮查看更多的设置（类似其他下载软件），但是我又不想做的太复杂了增加操作，并且我自己做界面就不太会。都是慢慢摸索，所以会改来改去。

像你最开始的思路就是点击选项后打开选项页面，然后关闭后返回这种，理论上是能做到。但是原则就是尽量减少操作的前提下，最开始的方案就是那个下滑功能，然后改成了现在这种直接把窗口变大。 OTL

**Reply by wordpure** [reply]
> 我现在主要是想默认打开新下载就包含了所有选项，而抓取下载这种则是个简化窗口，但是用户能通过点击按钮查看更多的设置（类似其他下载软件

还真没注意到这点😂

右键菜单弹出窗口最大的问题，就是点击选项后窗口变化不明显，还有计时器不会停止，其他的倒没什么。

不过像下载文件夹、用户代理这两个，修改的频率比较高，可能默认显示会比较好。

**Reply by jc3213** [reply]
这样？
![before](https://user-images.githubusercontent.com/8744936/201465007-b4ad70e3-954b-452b-8719-554d1ed24f31.png)
![after](https://user-images.githubusercontent.com/8744936/201465008-a6ce821e-50a7-43d0-bb5e-038f64d12b5c.png)

**Reply by wordpure** [reply]
> 这样？ ![before](https://user-images.githubusercontent.com/8744936/201465007-b4ad70e3-954b-452b-8719-554d1ed24f31.png) ![after](https://user-images.githubusercontent.com/8744936/201465008-a6ce821e-50a7-43d0-bb5e-038f64d12b5c.png)

下载文件夹和文件名的位置能否对调下？

**Comment by jc3213** [comment]
帮忙测试一下最新的`main`分支吧，看看有啥问题？

**Reply by wordpure** [reply]
> 帮忙测试一下最新的`main`分支吧，看看有啥问题？

就之前提到的 [这两点](https://github.com/jc3213/download_with_aria2/discussions/45#discussioncomment-4122286)。

**Reply by jc3213** [reply]
数据验证这个我是真不想做，这样就必须每个输入窗口单独写逻辑了。你只能说`<input>`的各个类别`type`并没有内建数据验证机制。专门去做这个机制太麻烦了。

然后抓取获取文件名跟右键菜单这个，`contextMenus`的onClicked事件并没有提供任何跟文件名相关的事项，虽然可以通过content-script进行互动来读取类似`downoad`属性之类的附带文件名，但是大多数情况下右键菜单是无法获取到文件名的。除非修改逻辑为正常访问再抓取，或者直接通过`fetch`之类的API来操作。

**Reply by wordpure** [reply]
原来是这样😂

其他方面没看到什么问题。

**Reply by jc3213** [reply]
都2022年了，浏览器还没内建数据验证我也觉得挺离谱的，如果只是1个，2个输入框，写一写就算了。主要还是太复杂了。太多东西要做匹配，而且匹配规则还不一样。

我也是看Aria2 WebUI跟YAAW都没做数据验证，我也就懒得弄了。他们那种都没做，我一个比他们还小的程序，那还是躺平吧。

然后就是并不是所有anchor都附带了`download`属性，大多数都没有，有些`download`属性并不是文件正确的名字，这里实在太混乱了，我个人并不想做这个适配。

倒是想做其他下载软件那种，按住Alt点anchor然后自动发送到aria2，不需要右键contextMenu，但是这里的逻辑交互感觉很麻烦，必须做全局`click`事件监听或者直接`document.querySelectorAll`选择所有anchor，感觉哪种方案都挺蠢的（毕竟拖慢性能）

**Comment by jc3213** [comment]
发现忘了做`download_headers`的逻辑了，大部分内容要改写（其实又写回2.xxxx版本了），目前仅测试了Chromium跟Chromium MV3版本，Firefox回头测试

**Comment by jc3213** [comment]
- Fixed a bug that click on `New` on `Task Manager` won't open `New Download` window
- Fixed error key mapping on `Options`
- Fixed indicator for `Manifest V3`
- Changed `Save` logic on `Options` that not to clear changes history
- Implemented `Download Headers` logic that is forgotten
- Update UI
- Update i18n

<hr /><em>This discussion was created from the release <a href='https://github.com/jc3213/download_with_aria2/releases/tag/4.0.0b6'>4.0.0 Beta6</a>.</em>
